Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,296 @@
+                              advertisedChecksum == actualChecksum
+                        else {
+                            throw RegistryError.invalidChecksum(
+                                expected: expectedChecksum?.hexadecimalRepresentation ?? advertisedChecksum,
+                                actual: actualChecksum
+                            )
+                        }
+
+                        let archivePath = path.withExtension("zip")
+                        try fileSystem.writeFileContents(archivePath, bytes: contents)
+
+                        try fileSystem.createDirectory(path, recursive: true)
+
+                        let archiver = ZipArchiver(fileSystem: fileSystem)
+                        archiver.extract(from: archivePath, to: path) { result in
+                            try? fileSystem.removeFileTree(archivePath)
+                            completion(result)
+                        }
+                    } catch {
+                        try? fileSystem.removeFileTree(path)
+                        throw error
+                    }
+                } else {
+                    throw RegistryError.invalidResponse
+                }
+            })
+        }
+    }
+}
+
+private func loadManifest(
+    _ contents: ByteString,
+    baseURL: URL,
+    manifestLoader: ManifestLoaderProtocol,
+    toolsVersion: ToolsVersion,
+    swiftLanguageVersion: SwiftLanguageVersion? = nil,
+    packageKind: PackageReference.Kind = .local
+) throws -> Manifest {
+    let fs = InMemoryFileSystem()
+
+    let filename: String
+    if let swiftLanguageVersion = swiftLanguageVersion {
+        filename = Manifest.basename + "@swift-\(swiftLanguageVersion).swift"
+    } else {
+        filename = Manifest.basename + ".swift"
+    }
+
+    try fs.writeFileContents(AbsolutePath.root.appending(component: filename), bytes: contents)
+    return try manifestLoader.load(
+        package: AbsolutePath.root,
+        baseURL: baseURL.lastPathComponent,
+        toolsVersion: toolsVersion,
+        packageKind: packageKind,
+        fileSystem: fs
+    )
+}
+
+private extension String {
+    /// Drops the given suffix from the string, if present.
+    func spm_dropPrefix(_ prefix: String) -> String {
+        if hasPrefix(prefix) {
+            return String(dropFirst(prefix.count))
+        }
+        return self
+    }
+}
+
+private extension AbsolutePath {
+    func withExtension(_ extension: String) -> AbsolutePath {
+        guard !self.isRoot else { return self }
+        let `extension` = `extension`.spm_dropPrefix(".")
+        return AbsolutePath(self, RelativePath("..")).appending(component: "\(basename).\(`extension`)")
+    }
+}
+
+private extension ByteString {
+    var sha256Checksum: ByteString {
+        #if canImport(CryptoKit)
+        if #available(macOS 10.15, *) {
+            return CryptoKitSHA256().hash(self)
+        }
+        #endif
+
+        return SHA256().hash(self)
+    }
+}
```

Comment:
there are all very useful and can move to Basics

Reply:
See #3116